gapic/java/java_gapic_pkg_deps_resolution.bzl

```diff
@@ -0,0 +1,190 @@
+    corresponding jar dependency file's full path. This implies that maven_jar targets are named
+    following the convention documented in bazel's official documentation (a relatively safe
+    assumption).
+
+    The aforementioned bazel documentation link: https://docs.bazel.build/versions/master/be/workspace.html#maven_jar
+
+    The implementation approach is dictated by the fact that maven_jar rule and corresponding java_*
+    bazel rules drop the original artifact information (after downloading the jar is completely
+    "demavenized"). Also there is really no good way of retreiving the value of the "artifact"
+    attribute value (passed to maven_jar rule) during analysis phase (note, native.existing_rule()
+    is not available during analysis phase).
+
+    There is no support for regular expressions in Starlark, so all the string shenanigans are done
+    the "old school" way.
+
+    Args:
+        file: (File) The File object, representing the jar dependency. Required.
+        group_overrides: {"name": "group"} dictionary to explicitly override calculated group for an
+            artifact, for cases, when either convention (maven artifact naming or bazel repository
+            target naming) is not followed.
+
+    Returns:
+        tuple: (string, string, string, string) The artifact id in the form of a tuple
+            (group, name, version, classifier), for example ("com.google.api", "gax-grpc", "1.30.0", "testlib")
+    """
+
+    # Reconstructing the `name:version:classifier` portion from basename
+    dirname = file.dirname
+    extension = file.extension
+    basename = file.basename[:-len(extension) - 1]
+
+    if basename.endswith("-ijar") or basename.endswith("-hjar"):
+        basename = basename[:-len("-ijar")]
+
+    # Representing [group, name, version, classifier]
+    artifact_id = ["", "", "", ""]
+    # Meven splits artifact name version and classifier with '-'
+    chunks = basename.split("-")
+    for i in range(0, len(chunks)):
+        if not artifact_id[1]:
+            # Find first chunk, which contains only digits and dots, recognize this as the beginning
+            # of the version portion of artifact id.
+            if _possibly_artifact_id_version_chunk(chunks[i]):
+                artifact_id[1] = "-".join(chunks[0:i])
+                artifact_id[2] = chunks[i]
+        elif not artifact_id[3]:
+            # Everything, which comes after version chunk and does not have digits in it is
+            # recognized as a classifier chunk, otherwise (if there is at least one digit) it is
+            # recognized as a continuation of the version and appended to the artifact version
+            # portion accordingly.
+            if _possibly_artifact_id_classifier_chunk(chunks[i]):
+                artifact_id[3] = "-".join(chunks[i: len(chunks)])
+                break
+            else:
+                artifact_id[2] = "-".join([artifact_id[2], chunks[i]])
```

Comment:
Is there any way to write unit tests for starlark code? This seems heavy on raw logic, which is something I feel has the most value for being tested.

Reply:
There is a way to test starlark rules: https://docs.bazel.build/versions/master/skylark/testing.html, but please lets not do it now. As for this particular logic, I believe the proper way of reducing complexity here would be to request from bazel team to have `maven_jar` rule to return a provider with the actual artifact_id value in t (i.e. to not drop maven-specific info too early). This would allow to get rid from 90% of the `java_gapic_pkg_deps_resolution.bzl` file and also remove all the heuristics. Unless i'm missing something, this should be easy to implement (the only potential problem is that bazel team may not want to keep that information for some reason).

Reply:
Alright - let's do at least one of those things (add tests or get Bazel to return the data from maven_jar that we want); I don't want to have complex, untested rule code indefinitely.